### ISSUE TITLE
ENH: Make approx_derivative for finite differene gradients public

### ISSUE THREAD

**Issue by dschmitz89** [scipy/scipy#14946]
### Is your feature request related to a problem? Please describe.

The function [approx_derivative](https://github.com/scipy/scipy/blob/v1.6.3/scipy/optimize/_numdiff.py) is a powerful function for finite difference gradients and also respects parameter bounds. All scipy's minimizers already use it internally, I also use it in my [simplenlopt project](https://github.com/dschmitz89/simplenlopt). Now that private functions will be deprecated, I would like to start a discussion if this function could become public and if yes what would be needed. I find it very convenient already, only bounds are defined differently than in the rest of the optimize module.

### Describe the solution you'd like.

A public function approx_derivative with the same functionality as now.

**Comment by rgommers** [comment]
Good question @dschmitz89. `approx_derivative` is already very well documented, and the request seems reasonable to me. Part of the history here is that for a long time we thought about adding a `scipy.diff` submodule for finite differences, but that never materialized. I'm not sure if we ever discussed exposing `approx_derivative` in the `optimize` namespace - I searched the mailing list, there's lots of discussion about the behavior of this function, but nothing about making it public.

Let's see what others think. @andyfaff, @ev-br, @mdhaber, @nmayorov, or anyone else?

**Comment by andyfaff** [comment]
W.r.t public + private functions I've been working under the assumption that if a function wasn't mentioned in a modules/files `__all__`, or if it wasn't exported from a file prefixed with `_` (the case here), then it was private. Private functions are subject to change, so use at own risk. It's been useful having `approx_derivative` as a private function as this has enabled its improvement (it's a core function supporting all the minimisers) without having to worry about back compatibility for public usage. Once public that flexibility is lost. Having said that, it's reasonably mature, and I support adding this to the public namespace.

I would suggest changing the `bounds` keyword to prefer a new-style `Bounds` object, which is used by `minimize`; a tuple of (lower, upper) bounds is almost a third way of specifying bounds.
This would involve changing the docstring of the function to be:
```
bounds : scipy.optimize.Bounds, optional
        Lower and upper bounds on independent variables. Defaults to no bounds.
        Use it to limit the range of function evaluation. Bounds checking is not
        implemented when `as_linear_operator` is True.
```
The function signature would change to:
```
bounds=None
```
Inside `_prepare_bounds` the relevant change would be:
```
if bounds is None:
    lb = np.full_like(x0, -np.inf, dtype=float)
    ub = np.full_like(x0, np.inf, dtype=float)
    return lb, ub
elif isinstance(bounds, scipy.optimize.Bounds):
    lb, ub = bounds.lb, bounds.ub
else:
    # to retain back compatibility
    # possible DeprecationWarning?
    lb, ub = [np.asarray(b, dtype=float) for b in bounds]
```

**Comment by rgommers** [comment]
> W.r.t public + private functions I've been working under the assumption that if a function wasn't mentioned in a modules/files `__all__`, or if it wasn't exported from a file prefixed with `_` (the case here), then it was private. Private functions are subject to change, so use at own risk. It's been useful having `approx_derivative` as a private function as this has enabled its improvement (it's a core function supporting all the minimisers) without having to worry about back compatibility for public usage. Once public that flexibility is lost.

Just to confirm: that all sounds right to me.

**Comment by andyfaff** [comment]
#11158 is related, as there are some changes in there that might make sense to apply before approx_derivative was made public.

**Comment by ev-br** [comment]
Retrospectively, one issue with `approx_derivative` is that it lumps up computation of derivatives, gradients and jacobians into a single function. This way, it cannot be used e.g. in stats (and elsewhere) for derivatives of scalar functions with broadcasting.
Having said that, no improvement has materialized so far, and `approx_derivative` is definitely useful. 

Overall, I'm +1 for making approx_derivative public, with tweaks mentioned by @andyfaff : the new-style Bounds and finishing up gh-11158.

**Comment by dschmitz89** [comment]
Is the internal `approx_derivative` function supposed to be superseded by the new `differentiate` module? I did not follow those developments closely. Especially important is if the new module supports bounds. CC @andyfaff @mdhaber

**Comment by andyfaff** [comment]
At the moment there's no plans to swap from approx_derivative for various minimize uses.its tuned to get the minimum number of function evals for numerical differentiation. For this issue it may be better to point users to differentiate.

**Comment by mdhaber** [comment]
> Is the internal `approx_derivative` function supposed to be superseded by the new `differentiate` module

No, probably not. `approx_derivative` is going to be faster, and the accuracy is fine for use in optimization.

> One crucial question is for example if the new module supports bounds

It does support one-sided finite differences, so it seems that it could support bounds if that is the key feature.

**Comment by andyfaff** [comment]
I think if there's feature parity then there's not much drive to make `approx_derivative` public. The only remaining factor would be the speed.

What about deprecating `approx_fprime` (and associated functions), and making `approx_derivative` public in `scipy.differentiate`?

> It only supports objectives returning Python scalars. (approx_derivative)

That's not correct, approx_derivative differentiates $R^n \rightarrow R^n$, so it can differentiate functions accepting and returning vectors.

**Comment by ev-br** [comment]
> approx_derivative differentiates R^n -> R^n, so it can differentiate functions accepting and returning vectors.

If anything, a design decision which looks suboptimal in hindsight was to make it automatically switch between computing derivatives of scalar functions and jacobians based on the dimensionality of inputs. This makes it impossible to for instance handle batches of scalar functions, which `stats.misc.derivative` was doing. 
I'm not current on what `scipy.differentiate` does, I only hope it does not make the same mis-step.

**Comment by dschmitz89** [comment]
Closing as there is now the much more powerful differentiate submodule.
